File: src/content/docs/topics/cs/leetcode-150/linked-list/002-add-two-numbers.py
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out

def from_list(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next

def add_two_numbers(l1, l2):
    pass  # TODO: implement

def _run_tests():
    assert to_list(add_two_numbers(from_list([2, 4, 3]), from_list([5, 6, 4]))) == [7, 0, 8]
    assert to_list(add_two_numbers(from_list([0]), from_list([0]))) == [0]
    assert to_list(add_two_numbers(from_list([9, 9, 9, 9, 9, 9, 9]), from_list([9, 9, 9, 9]))) == [8, 9, 9, 9, 0, 0, 0, 1]
    assert to_list(add_two_numbers(from_list([1]), from_list([2]))) == [3]
    assert to_list(add_two_numbers(from_list([5]), from_list([5]))) == [0, 1]
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/linked-list/002-add-two-numbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// TODO: implement
	return nil
}

func makeList(vals []int) *ListNode {
	if len(vals) == 0 {
		return nil
	}
	head := &ListNode{Val: vals[0]}
	cur := head
	for _, v := range vals[1:] {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return head
}

func listToSlice(head *ListNode) []int {
	var result []int
	for head != nil {
		result = append(result, head.Val)
		head = head.Next
	}
	return result
}

func slicesEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func runTests() {
	assert(slicesEqual(listToSlice(addTwoNumbers(makeList([]int{2, 4, 3}), makeList([]int{5, 6, 4}))), []int{7, 0, 8}))
	assert(slicesEqual(listToSlice(addTwoNumbers(makeList([]int{0}), makeList([]int{0}))), []int{0}))
	assert(slicesEqual(listToSlice(addTwoNumbers(makeList([]int{9, 9, 9, 9, 9, 9, 9}), makeList([]int{9, 9, 9, 9}))), []int{8, 9, 9, 9, 0, 0, 0, 1}))
	assert(slicesEqual(listToSlice(addTwoNumbers(makeList([]int{1}), makeList([]int{2}))), []int{3}))
	assert(slicesEqual(listToSlice(addTwoNumbers(makeList([]int{5}), makeList([]int{5}))), []int{0, 1}))
	fmt.Println("all tests pass")
}

func main() { runTests() }
